### PR TITLE
Implement the authentication flow

### PR DESCRIPTION
**Pull request by dtinth**
As specified by #5 

We implement the flow first (doing it top-down) so that we can unit-test all the edge cases that I can think of.

### CODE REVIEW

File: authenticationFlow.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    if (!result.idToken) {
+      return { error: 'Invalid password' }
+    }
+    return { idToken: result.idToken }
+  }
+}
+
+exports.signUp =
+function * signUp (username, email, password, {
+  log = (message) => console.log('[loginByUsernamePassword]', message),
```
Comment (dtinth) [inline authenticationFlow.js]: Wrong log

File: __tests__/authenticationFlow.test.js [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+      it('should be prevented', () => {
+        const env = createEnv()
+        env.givenLegacyUser('THAI', 'thai@bemuse.ninja', 'wow')
+        env.givenPlayer('THAI')
+        env.externalSignUp('player0', 'player0@bemuse.ninja', 'strongpassword').mustFail()
+        return env.verify()
+      })
+    })
+    describe('an unlinked account (somehow)', () => {
+      it('should try to link with legacy user via email, if possible')
```
Comment (dtinth) [inline __tests__/authenticationFlow.test.js]: Nope, an unlinked account can’t be created in the first place if its email clashes with a legacy user’s email.

File: __tests__/authenticationFlow.test.js [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    describe('hijacking user by creating account with existing player id as username', () => {
+      it('should be prevented', () => {
+        const env = createEnv()
+        env.givenLegacyUser('THAI', 'thai@bemuse.ninja', 'wow')
+        env.givenPlayer('THAI')
+        env.externalSignUp('player0', 'player0@bemuse.ninja', 'strongpassword').mustFail()
+        return env.verify()
+      })
+    })
+    describe('an unlinked account (somehow)', () => {
```
Comment (dtinth) [inline __tests__/authenticationFlow.test.js]: Example of unlinked accounts:

- Twitter sign-in: User signs in via Twitter and gets an OAuth key. Since this account has not been pre-registered, it is not linked, and must be linked manually.

File: __tests__/authenticationFlow.test.js [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    databaseGetUser (username) {
+      const user = legacyUserByEmail[username] || (() => {
+        const player = playerById[username]
+        return player && legacyUserByUsername[player.playerName]
+      })()
+      return user
+    },
+    rule (user) {
+      if (user.email && !user.playerName) {
+        const player = link(user._id, user.username)
```
Comment (dtinth) [inline __tests__/authenticationFlow.test.js]: How about linking at the client side?

This allows user-initiated linking (such as from Twitter) at the client-side.

File: __tests__/authenticationFlow.test.js [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    },
+    databaseGetUser (username) {
+      const user = legacyUserByEmail[username] || (() => {
+        const player = playerById[username]
+        return player && legacyUserByUsername[player.playerName]
+      })()
+      return user
+    },
+    rule (user) {
+      if (user.email && !user.playerName) {
```
Comment (dtinth) [inline __tests__/authenticationFlow.test.js]: `&& user.username`
